File: app/ui/src/app/models/coupon/coupon.ts
import { User } from "../user/user";

export interface Coupon {
    id?: number;              // Optional, for newly created coupons
    code: string;             // Unique coupon code
    percentage: number;       // Discount percentage
    totalUses: number;        // Total uses of this coupon
    expiryDate: string;       // Expiry date in ISO format
    users?: Set<User>;      // Users who have used this coupon (user IDs)
    payments?: Array<number>; // Payments where this coupon was applied (payment IDs)
    active: boolean;          // Indicates if the coupon is active
}
export interface CouponResponse {
    valid: boolean;         // Indicates if the coupon is valid
    discount?: number;      // Discount percentage if valid
    message: string;          // Message providing information about the coupon validity
}

File: app/backend/src/main/java/com/dcg/mvc/coupon/Coupon.java
package com.dcg.mvc.coupon;

import com.dcg.common.BaseEntity;
import com.dcg.mvc.payment.Payment;
import com.dcg.mvc.user.User;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;

@Getter
@Setter
@SuperBuilder
@AllArgsConstructor
@NoArgsConstructor
@Entity
@ToString
@Table(name = "coupon")
public class Coupon extends BaseEntity {


    @Column(nullable = false, unique = true)
    private String code; // Unique coupon code

    @Column(nullable = false)
    private double percentage; // Discount percentage

    @Column(nullable = false)
    private int totalUses; // Total uses of this coupon

    @Column(nullable = false)
    private LocalDateTime expiryDate; // Expiry date for the coupon

    @ManyToMany(mappedBy = "couponsUsed")
    private Set<User> users; // Users who have used this coupon

    @OneToMany(mappedBy = "coupon")
    @JsonIgnore
    private List<Payment> payments; // Payments where this coupon was applied

    private boolean active;

    public boolean isValid() {
        return totalUses > 0 && expiryDate.isAfter(LocalDateTime.now());
    }

    public void decrementUses() {
        if (totalUses > 0) {
            totalUses--;
        }
    }
}
